File: src/main/java/com/gluck/jobtracker/exception/GlobalExceptionHandler.java
package com.gluck.jobtracker.exception;

import com.gluck.jobtracker.model.ApiErrorResponse;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;

import java.util.HashMap;
import java.util.Map;

@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<Map<String, String>> handleValidationExceptions(MethodArgumentNotValidException ex) {
        Map<String, String> info = new HashMap<>();
        var errors = ex.getBindingResult();
        errors.getFieldErrors().forEach(error ->
                info.put(error.getField(), error.getDefaultMessage() != null ? error.getDefaultMessage() : "Invalid field")
        );
        return new ResponseEntity<>(info, HttpStatus.BAD_REQUEST);
    }

    @ExceptionHandler(NoSuchJobFoundException.class)
    public ResponseEntity<ApiErrorResponse> handleNoSuchJobException(NoSuchJobFoundException ex) {
        ApiErrorResponse errorResponse = new ApiErrorResponse(ex.getMessage() != null ? ex.getMessage() : "Unknown error");
        return new ResponseEntity<>(errorResponse, HttpStatus.NOT_FOUND);
    }

}

File: src/main/kotlin/com/gluck/jobtracker/exception/GlobalExceptionHandler.kt
package com.gluck.jobtracker.exception

import com.gluck.jobtracker.model.ApiErrorResponse
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class GlobalExceptionHandler {

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationExceptions(ex: MethodArgumentNotValidException): ResponseEntity<Map<String, String>> {

        val info = mutableMapOf<String, String>()
        val errors = ex.bindingResult
        for (error in errors.fieldErrors) {
            info[error.field] = error.defaultMessage ?: "Invalid field"
        }
        return ResponseEntity(info, HttpStatus.BAD_REQUEST)

    }

    @ExceptionHandler(NoSuchJobFoundException::class)
    fun handleNoSuchJobException(ex: NoSuchJobFoundException): ResponseEntity<ApiErrorResponse> {

        val errorResponse = ApiErrorResponse(ex.message ?: "Unknown error")
        return ResponseEntity(errorResponse, HttpStatus.NOT_FOUND)

    }

}